Pulseaudio loopback switch

Cinntax opened the pull request:
**Description:**
A new platform for the switch module to allow for loading/unloading pulseaudio loopback modules.  This is to support a whole home audio system.

**Related issue (if applicable):** #
N/A

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
switch Loopback3:
  name: Airplay -> Main Level
  platform: pulseaudio_loopback
  sink_name: main_level
  source_name: AirplayInput1.monitor
```

**Checklist:**
- [X] Local tests with `tox` run successfully.
- [X] TravisCI does not fail. **Your PR cannot be merged unless CI is green!**
- [X] [Fork is up to date](http://stackoverflow.com/a/7244456) and was rebased on the `dev` branch before creating the PR.
- [X] Commits have been [squashed](https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit).
- If code communicates with devices:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
  - [X] New dependencies are only imported inside functions that use them ([example](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
  - [X] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.
- If the code does not interact with devices:
  - [] Tests have been added to verify that the new code works.

Cinntax committed:
Creation of a new platform for the existing switch component.

balloob commented on homeassistant/components/light/__init__.py:
```diff
@@ -227,7 +227,7 @@ def handle_light_service(service):
 
         if ATTR_COLOR_TEMP in dat:
-            # color_temp should be an int of mirads value
+            # color_temp should be an int of mireds value
```
This is not part of your PR.

balloob commented on homeassistant/components/switch/pulseaudio_loopback.py:
```diff
@@ -0,0 +1,136 @@
+
+    def update(self):
+        """Refresh state in case an alternate process modified this data."""
+        import re
```
You can import `re` at the top.

balloob commented on homeassistant/components/switch/pulseaudio_loopback.py:
```diff
@@ -0,0 +1,136 @@
+        if result:
+            self._module_id = result.group(1)
+        else:
+            self._module_id = "-1"
```
Why set it to a string if you only ever use it as an int?

Cinntax replied:
I had considered making it int... when interfacing with pulseaudio, I need to make it a string again, so it was just a matter of when I need to cast it.  If it was confusing to you, it will be to others too- so I'll adjust.

balloob commented on homeassistant/components/switch/pulseaudio_loopback.py:
```diff
@@ -0,0 +1,136 @@
+        self.update_ha_state()
+
+    @property
+    def should_poll(self):
```
Default is `True` so no need to add property.

Cinntax commented:
Thanks! I'll get these resolved.  In terms of the "re" import- should I move the socket import to the top as well?  Are you recommending this because they're standard python modules?  Just trying to understand the rules.

balloob commented:
Yeah, only 3rd party dependencies that home Assistant will install for you
need to be imported at function level

On Sun, Mar 20, 2016, 16:09 Dan Cinnamon <notifications@github.com> wrote:

> Thanks! I'll get these resolved. In terms of the "re" import- should I
> move the socket import to the top as well? Are you recommending this
> because they're standard python modules? Just trying to understand the
> rules.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/1583#issuecomment-199057101>
>

Cinntax commented:
Okay- I've made the following changes:
1- I've moved my (python standard) imports back up to the top of the file.
2- I got rid of the "shouldpoll" property, since it's not necessary.
3- I took out the stray commit that didn't belong there.
4- I adjusted the module_id variable, calling it the more appropriate module_idx instead, and making it an integer.
Thanks!

balloob commented:
Looks great :dolphin:
